akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -432,27 +432,45 @@ import scala.util.control.NonFatal
     override def initialAttributes: Attributes = DefaultAttributes.futureSource
     override def createLogic(attr: Attributes) =
       new GraphStageLogic(shape) with OutHandler {
+        // Always register a callback which will fail the stage if the future fails...
+        val cb = getAsyncCallback[Try[T]](onFutureCompleted).invoke _
+        future.onComplete(cb)(ExecutionContexts.parasitic)
```

Comment:
Extract the code in `onPull` to a dedicated method , and then we can keep the optimization if the future is already completed.